### PR TITLE
feat(TestScheduler): add support for testing promises synchronously

### PR DESCRIPTION
**Pull request by staltz**
**DO NOT MERGE.**

FAILED attempt at adding support for testing promises synchronously with a VirtualTimeScheduler. This
commit is just to illustrate the attempt and highlight how exactly does it not work.

For issue #701.

- - -

After spending some time punching Promises in the face trying to make them behave synchronously like Observables, I gave up on this endeavor because I'm fairly convinced it's not possible. Here is one simple example:

```js
describe('debounce', function () {
  it('should accept a Promise as the duration selector ', function () {
    var e1 =    hot('--a--bc--d----|');
    var ptemplate =   '--|          ';
    var expected =  '----a---c--d--|';

    var result = e1.debounce(function durationSelector() {
      return promise(ptemplate, 42);
    });

    expectObservable(result).toBe(expected);
  });
```

And when executed:
```
Failures:
1) Observable.prototype.debounce() should accept a Promise as the duration selector 
  Message:
    Expected 
    {"frame":140,"notification":{"kind":"N","value":"d","hasValue":true}}
    {"frame":140,"notification":{"kind":"C","hasValue":false}}
    
    to deep equal 
    {"frame":40,"notification":{"kind":"N","value":"a","hasValue":true}}
    {"frame":80,"notification":{"kind":"N","value":"c","hasValue":true}}
    {"frame":110,"notification":{"kind":"N","value":"d","hasValue":true}}
    {"frame":140,"notification":{"kind":"C","hasValue":false}}
 ```

In other words this means: **because Promises call `then()` at the next tick**, which happens **after** the rxTestScheduler has passed through all those frames, the promise as the duration selector for debounce will debounce the events for the next tick, **and not** for frames 40, 80, 110, as we specified. There is not much we can do since this is due to debounce's implementation, not related to the TestPromise helper I made.

I quit the attempt to support marble diagrams for Promises because it seems the ROI won't be positive. This PR is just to report my activity. I'll close this PR and issue #701 too.

CC @blesh @kwonoj

### CODE REVIEW

**Comment by SLaks** [comment]
Can this be implemented using `await` instead?
